torch/csrc/jit/codegen/cuda/interface.cpp

```diff
@@ -24,6 +24,103 @@ namespace cuda {
 
 static std::atomic<bool> cuda_fusion_guard_mode{true};
 
+// There are 3 sources of information on whether to enable nvfuser:
+// 1. assigned value from setEnabled() - takes precendence if it has been set
+// 2. value from environment variable - only used if setEnabled() is unset
+// 3. default value - used if both 1 and 2 are unset.
+//
+// If 1 or 2 tries to enable nvfuser when it cannot be enabled (e.g. cuda not
+// available), then an error will be thrown. The default will not error.
+class NVFuserEnabler {
+ private:
+  c10::optional<bool> runtime_assigned_fuser_enabled_ = c10::nullopt;
+  std::once_flag enabled_check_flag_;
+  std::mutex mutex_;
+
+  static bool nvfuserCanBeEnabled() {
+#ifdef USE_ROCM
+    bool has_rocm = true;
+#else
+    bool has_rocm = false;
+#endif
+    return at::globalContext().hasCUDA() && !has_rocm &&
+        NVFuserPassManager::isRegistered();
+  }
+
+  static void assertFuserCanBeEnabled(bool is_enabled) {
+    if (!is_enabled) {
+      return;
+    }
+    TORCH_CHECK(
+        nvfuserCanBeEnabled(),
+        "Running CUDA fuser is only supported on CUDA builds.");
+  }
+
+  static c10::optional<bool> getFuserEnabledEnvVar() {
+    static const char* enable_c_str = std::getenv("PYTORCH_JIT_ENABLE_NVFUSER");
+    if (!enable_c_str) {
+      return c10::nullopt;
+    }
+    std::string enable(enable_c_str);
+    if (enable == "0" || enable == "OFF") {
+      return false;
+    }
+    return true;
+  }
+
+  static c10::optional<bool> getCachedFuserEnabledEnvVar() {
+    static c10::optional<bool> default_enabled = getFuserEnabledEnvVar();
```

Comment:
NITPICK: since we cache the env variable, we might want to add a test to validate the env variable is NOT cached during initialization so the workflow below would work vvv
```
import os
import torch

os.environ['PYTORCH_JIT_ENABLE_NVFUSER'] = '1'
# running to check if nvfuser is enabled...
```

Reply:
> I think it would be cached during initialization... you could use torch._C._jit_set_nvfuser_enabled(True) though for this use case?

I don't have a real objection on this one.
We unfortunately have been using env_variable for some debugging knobs (back in the old days it's hard to push new APIs to PyTorch core) and have been using `os.environ['...'] = '...'` in our script. We've seen users picking up that pattern as well and cached value during initialization might hit people as a surprise.
In this case we do have the proper _C API, so we'll just stick to that in our scripts.

We should also go back and fix the other env_variable hack we have

Reply:
`os.environ['PYTORCH_JIT_LOG_LEVEL'] = ...` doesn't work either, in addition to NNC not working. It would be nice to update the env variables to not cache but as @davidberard98 notes this doesn't work for the other stuff. Maybe this is fine as is in this PR and then in other prs we can update them ?

The environment variables that control optimization passes it's fine to update, but we don't want to check environment variables on e.g. every running of the kernels